**Evaluation of comments**                   
Aim 1: Does Ask HN or Show HN recieve more comments?                       
Aim 2: At what time does posts recieve more comments on average?

In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0]
hn = hn[1:]

print(headers)
print()
print(hn[7])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55']


Make 3 lists and categorise the ask, show and other "titles" into these lists. Making sure to convert the titles to lowercase.        
Count the number of each in each list at the end


In [2]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title_lowercase = title.lower()
    
    if title_lowercase.startswith('ask hn'):
        ask_posts.append(row)
    elif title_lowercase.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print("Example from ask list:")        
print(ask_posts[0])
print()
# print(other_posts[:5])

num_ask = len(ask_posts)
print("We have {num} of ask posts.".format(num=num_ask))
num_show = len(show_posts)
print("We have {num} of show posts.".format(num=num_show))
num_other = len(other_posts)
print("We have {num} of other posts.".format(num=num_other))


Example from ask list:
['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55']

We have 1744 of ask posts.
We have 1162 of show posts.
We have 17194 of other posts.


Do ask posts or show posts receive more **comments** on average?

In [3]:
# Number of average comments for each ask_post
total_ask_comments = 0

for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
    
avg_ask_comments = total_ask_comments / num_ask

# print(avg_ask_comments)
    
# Number of average comments for each show_post
total_show_comments = 0

for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
    
avg_show_comments = total_show_comments / num_show

# print(avg_show_comments)

# Number of average comments for each other_post
total_other_comments = 0

for row in other_posts:
    num_comments = int(row[4])
    total_other_comments += num_comments
    
avg_other_comments = total_other_comments / num_other

# print(avg_other_comments)

print("We have an average of {num:.2f} comments within ask posts.".format(num=avg_ask_comments))
num_show = len(show_posts)
print("We have an average of {num:.2f} comments within show posts.".format(num=avg_show_comments))
num_other = len(other_posts)
print("We have an average of {num:.2f} comments within other posts.".format(num=avg_other_comments))

We have an average of 14.04 comments within ask posts.
We have an average of 10.32 comments within show posts.
We have an average of 26.87 comments within other posts.


**Findings**         
It appears that there are more comments on avergae with the ask posts than the show posts.

**Next Section**            
Just focusing on the ask_posts:
Do posts made at a certain time attract more posts?
1. Calculate the amount of posts in each hour of the day with the number of comments recieved
2. Calculate avergae number of comments ask posts receive by hour created

In [4]:
print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [5]:
from datetime import datetime as dt
result_list = []

# To separate the date and time created and the number of comments out into a new list
for row in ask_posts:
    created_at = row[6]
    num_comments = int(row[4])
    result_list.append([created_at, num_comments])
    
# To test
# print(ask_posts[:2]) 
# print()
print(result_list[:1])
print()    
    
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    num_comments = row[1]
    dt_string = row[0]
    dt_object1 = dt.strptime(dt_string, "%m/%d/%Y %H:%M") 
    hour = dt_object1.strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = num_comments
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += num_comments
    
    
print("Counts for each hour: ")
print(counts_by_hour)
print()
print("Comments for each hour: ")
print(comments_by_hour)

[['8/16/2016 9:55', 6]]

Counts for each hour: 
{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}

Comments for each hour: 
{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


**Calculating the average number of comments for posts during each hour of the day**


In [6]:
avg_by_hour = []

for key in counts_by_hour:
    avg = comments_by_hour[key] / counts_by_hour[key]
    avg_by_hour.append([key, round(avg, 2)])
    
# I need to sort the list     
avg_by_hour.sort()    

# To nicely print
for row in avg_by_hour:
    print(row)  

['00', 8.13]
['01', 11.38]
['02', 23.81]
['03', 7.8]
['04', 7.17]
['05', 10.09]
['06', 9.02]
['07', 7.85]
['08', 10.25]
['09', 5.58]
['10', 13.44]
['11', 11.05]
['12', 9.41]
['13', 14.74]
['14', 13.23]
['15', 38.59]
['16', 16.8]
['17', 11.46]
['18', 13.2]
['19', 10.8]
['20', 21.52]
['21', 16.01]
['22', 6.75]
['23', 7.99]


**Putting the highest average comments at the top of the list**

In [7]:
swap_avg_by_hour = []
for row in avg_by_hour:
    hour = row[0]
    avg = row[1]
    swap_avg_by_hour.append([avg, hour])

for row in swap_avg_by_hour:
    print(row)

[8.13, '00']
[11.38, '01']
[23.81, '02']
[7.8, '03']
[7.17, '04']
[10.09, '05']
[9.02, '06']
[7.85, '07']
[10.25, '08']
[5.58, '09']
[13.44, '10']
[11.05, '11']
[9.41, '12']
[14.74, '13']
[13.23, '14']
[38.59, '15']
[16.8, '16']
[11.46, '17']
[13.2, '18']
[10.8, '19']
[21.52, '20']
[16.01, '21']
[6.75, '22']
[7.99, '23']


In [8]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True)

for row in swap_avg_by_hour:
    print(row)

[8.13, '00']
[11.38, '01']
[23.81, '02']
[7.8, '03']
[7.17, '04']
[10.09, '05']
[9.02, '06']
[7.85, '07']
[10.25, '08']
[5.58, '09']
[13.44, '10']
[11.05, '11']
[9.41, '12']
[14.74, '13']
[13.23, '14']
[38.59, '15']
[16.8, '16']
[11.46, '17']
[13.2, '18']
[10.8, '19']
[21.52, '20']
[16.01, '21']
[6.75, '22']
[7.99, '23']


Aim is to make this format:               
15:00: 38.59 average comments per post

In [11]:
print("Top 5 Hours for Ask Posts Comments")

for row in sorted_swap[:5]:
    hour = dt.strptime(row[1], '%H')
    hour_str = hour.strftime('%H:%M')
    print("{time}: {avg} average comments per post".format(time= hour_str, avg= row[0]))
   

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.8 average comments per post
21:00: 16.01 average comments per post


**Final Recommendations:**                 
Above are the top five hours for a post that has the highest chance for the highest average comments. If you would like the maximium chances for a comment I would recommend posting at 2am, 3pm or 8pm. If I was to recommend one specific time, it appears that 3pm has the highest.